MinutesFromUTC has no tzname method

Class `pywbem.MinutesFromUTC ` is derived from `tzinfo` but does not implement the abstract method `tzname()`, resulting in NotImplementedError if a datetime object is asked to format its timezone.

Example:
```
import pywbem
from datetime import datetime

ti = pywbem.MinutesFromUTC(-720)
dt = datetime(1900, 6, 19, 1, 2, 3, tzinfo=ti)
print(dt.strftime('%Y-%m-%d %H:%M:%S %Z'))

Traceback (most recent call last):
  File "<string>", line 1, in <module>
NotImplementedError: a tzinfo subclass must implement tzname()
```

This was revealed via pylint on Python 3.7:
```
************* Module pywbem._cim_types
pywbem\_cim_types.py:169:0: W0223: Method 'tzname' is abstract in class 'tzinfo' but is not overridden (abstract-method)
```

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -34,6 +34,7 @@ requests>=2.20.1
 requests-mock>=1.6.0
 yagot>=0.5.0
 importlib-metadata<1,>=0.12; python_version <= '3.7'
+pytz>=2016.10
 
 # Unit test (indirect dependencies):
 # Pluggy 0.12.0 has a bug causing pytest plugins to fail loading on py38

--- a/docs/changes.rst
+++ b/docs/changes.rst
@@ -388,6 +388,10 @@ Released: not yet
 * Fixed version incompatibilities reported by pip for tox/pluggy,
   ipython/prompt-toolkit, and flake8/pyflakes. (See issue #2153)
 
+* Fixed the issue where formatting the timezone name of a pywbem.MinutesFromUTC
+  object raised NotImplementedError, by adding a tzname() method.
+  (see issue #2160)
+
 **Enhancements:**
 
 * Changed GetCentralInstances methodology in WBEMServer.get_central_instances()

--- a/minimum-constraints.txt
+++ b/minimum-constraints.txt
@@ -83,6 +83,7 @@ requests==2.20.1
 yagot==0.5.0
 requests-mock==1.6.0
 importlib-metadata==0.12; python_version <= '3.7'
+pytz==2016.10
 
 # Unit test (indirect dependencies):
 pluggy==0.7.1; python_version >= '2.7' and python_version <= '3.6'
@@ -195,7 +196,6 @@ pkginfo==1.4.1
 prompt-toolkit==2.0.1
 ptyprocess==0.5.1
 py==1.5.1
-pytz==2016.10
 pyzmq==16.0.4
 qtconsole==4.2.1
 requests-toolbelt==0.7.0

--- a/pywbem/_cim_types.py
+++ b/pywbem/_cim_types.py
@@ -254,6 +254,26 @@ class MinutesFromUTC(tzinfo):
         """
         return timedelta(0)
 
+    def tzname(self, dt):  # pylint: disable=unused-argument
+        """
+        An implementation of the corresponding base class method,
+        (see :meth:`py:datetime.tzinfo.tzname` for its description),
+        which needs to return the name of the timezone of the
+        specified datetime object.
+
+        This implementation returns the timezone offset formatted as a
+        signed HH:MM string, where positive values are east of UTC.
+        """
+        # Note that divmod() and // return one less for negative numbers
+        # assuming the remainder would be added on it. For example,
+        # divmod(-5, 60) = -1, 55
+        # That is mathematically consistent, but is not what we want
+        # since we want both return values to be negative, for negative
+        # input. Therefore we handle the sign separately.
+        sign = '-' if self._offset < 0 else ''
+        hh, mm = divmod(abs(self._offset), 60)
+        return "{sign}{hh:02d}:{mm:02d}".format(sign=sign, hh=hh, mm=mm)
+
 
 class CIMType(object):  # pylint: disable=too-few-public-methods
     """Base type for all CIM data types defined in this package."""

--- a/tests/unittest/pywbem/test_cim_types.py
+++ b/tests/unittest/pywbem/test_cim_types.py
@@ -8,6 +8,7 @@ from __future__ import absolute_import, print_function
 
 import re
 from datetime import timedelta, datetime
+import pytz
 import pytest
 import six
 
@@ -904,6 +905,380 @@ def test_datetime_str(datetime_init_tuple):
 # TODO: Add testcases for get_local_utcoffset()
 # TODO: Add testcases for now()
 # TODO: Add testcases for fromtimestamp()
+
+
+TESTCASES_MINUTESFROMUTC_INIT = [
+
+    # Testcases for MinutesFromUTC.__init__()
+
+    # Each list item is a testcase tuple with these items:
+    # * desc: Short testcase description.
+    # * kwargs: Keyword arguments for the test function:
+    #   * args: Positional args for init.
+    #   * kwargs: Keyword args for init.
+    #   * exp_offset: Expected offset to UTC in minutes.
+    # * exp_exc_types: Expected exception type(s), or None.
+    # * exp_warn_types: Expected warning type(s), or None.
+    # * condition: Boolean condition for testcase to run, or 'pdb' for debugger
+
+    (
+        "Test with missing required arguments",
+        dict(
+            args=[],
+            kwargs=dict(),
+            exp_offset=None,
+        ),
+        TypeError, None, True
+    ),
+    (
+        "Test with minimal number of positional arguments",
+        dict(
+            args=[30],
+            kwargs=dict(),
+            exp_offset=30,
+        ),
+        None, None, True
+    ),
+    (
+        "Test with minimal number of keyword arguments",
+        dict(
+            args=[],
+            kwargs=dict(
+                offset=30,
+            ),
+            exp_offset=30,
+        ),
+        None, None, True
+    ),
+    (
+        "Test with too many positional arguments",
+        dict(
+            args=[30, 0],
+            kwargs=dict(),
+            exp_offset=None,
+        ),
+        TypeError, None, True
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    "desc, kwargs, exp_exc_types, exp_warn_types, condition",
+    TESTCASES_MINUTESFROMUTC_INIT)
+@simplified_test_function
+def test_MinutesFromUTC_init(testcase, args, kwargs, exp_offset):
+    """
+    Test function for MinutesFromUTC.__init__().
+    """
+
+    # The code to be tested
+    tz = MinutesFromUTC(*args, **kwargs)
+
+    # Ensure that exceptions raised in the remainder of this function
+    # are not mistaken as expected exceptions
+    assert testcase.exp_exc_types is None
+
+    # Verify the UTC offset of the tz object.
+    exp_utcoffset = timedelta(minutes=exp_offset)
+    dt = datetime.now()
+    assert tz.utcoffset(dt) == exp_utcoffset
+
+
+TESTCASES_MINUTESFROMUTC_UTCOFFSET = [
+
+    # Testcases for MinutesFromUTC.utcoffset()
+
+    # Each list item is a testcase tuple with these items:
+    # * desc: Short testcase description.
+    # * kwargs: Keyword arguments for the test function:
+    #   * offset: offset argument for MinutesFromUTC().
+    #   * dt: datetime object as input for utcoffset().
+    #   * exp_utc_offset: Expected offset in minutes returned by utcoffset().
+    # * exp_exc_types: Expected exception type(s), or None.
+    # * exp_warn_types: Expected warning type(s), or None.
+    # * condition: Boolean condition for testcase to run, or 'pdb' for debugger
+
+    (
+        "Test with UTC offset 30 and datetime without tzinfo",
+        dict(
+            offset=30,
+            dt=datetime.now(),
+            exp_utc_offset=30,
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset -30 and datetime without tzinfo",
+        dict(
+            offset=-30,
+            dt=datetime.now(),
+            exp_utc_offset=-30,
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 150 and datetime with tzinfo for EST (ignored)",
+        dict(
+            offset=150,
+            dt=datetime.now(pytz.timezone('US/Eastern')),
+            exp_utc_offset=150,
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset -150 and datetime with tzinfo for EDT (ignored)",
+        dict(
+            offset=-150,
+            dt=pytz.timezone('US/Eastern').
+            localize(datetime.now(), is_dst=True),
+            exp_utc_offset=-150,
+        ),
+        None, None, True
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    "desc, kwargs, exp_exc_types, exp_warn_types, condition",
+    TESTCASES_MINUTESFROMUTC_UTCOFFSET)
+@simplified_test_function
+def test_MinutesFromUTC_utcoffset(testcase, offset, dt, exp_utc_offset):
+    """
+    Test function for MinutesFromUTC.utcoffset().
+    """
+
+    tz = MinutesFromUTC(offset)
+
+    # The code to be tested
+    act_utcoffset = tz.utcoffset(dt)
+
+    # Ensure that exceptions raised in the remainder of this function
+    # are not mistaken as expected exceptions
+    assert testcase.exp_exc_types is None
+
+    exp_utcoffset = timedelta(minutes=exp_utc_offset)
+    assert act_utcoffset == exp_utcoffset
+
+
+TESTCASES_MINUTESFROMUTC_DST = [
+
+    # Testcases for MinutesFromUTC.dst()
+
+    # Each list item is a testcase tuple with these items:
+    # * desc: Short testcase description.
+    # * kwargs: Keyword arguments for the test function:
+    #   * offset: offset argument for MinutesFromUTC().
+    #   * dt: datetime object as input for dst().
+    #   * exp_dst_offset: Expected DST offset in minutes returned by dst().
+    # * exp_exc_types: Expected exception type(s), or None.
+    # * exp_warn_types: Expected warning type(s), or None.
+    # * condition: Boolean condition for testcase to run, or 'pdb' for debugger
+
+    (
+        "Test with UTC offset 30 and datetime without tzinfo",
+        dict(
+            offset=30,
+            dt=datetime.now(),
+            exp_dst_offset=0,
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset -30 and datetime without tzinfo",
+        dict(
+            offset=-30,
+            dt=datetime.now(),
+            exp_dst_offset=0,
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 150 and datetime with tzinfo for EST (ignored)",
+        dict(
+            offset=150,
+            dt=datetime.now(pytz.timezone('US/Eastern')),
+            exp_dst_offset=0,
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset -150 and datetime with tzinfo for EDT (ignored)",
+        dict(
+            offset=-150,
+            dt=pytz.timezone('US/Eastern').
+            localize(datetime.now(), is_dst=True),
+            exp_dst_offset=0,
+        ),
+        None, None, True
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    "desc, kwargs, exp_exc_types, exp_warn_types, condition",
+    TESTCASES_MINUTESFROMUTC_DST)
+@simplified_test_function
+def test_MinutesFromUTC_dst(testcase, offset, dt, exp_dst_offset):
+    """
+    Test function for MinutesFromUTC.dst().
+    """
+
+    tz = MinutesFromUTC(offset)
+
+    # The code to be tested
+    act_dst = tz.dst(dt)
+
+    # Ensure that exceptions raised in the remainder of this function
+    # are not mistaken as expected exceptions
+    assert testcase.exp_exc_types is None
+
+    exp_dst = timedelta(minutes=exp_dst_offset)
+    assert act_dst == exp_dst
+
+    # Test the condition for sane tzinfo implementations w.r.t. DST
+    dt1 = datetime(2020, 4, 11, 13, 14, 15, tzinfo=tz)
+    diff1 = tz.utcoffset(dt1) - tz.dst(dt1)
+    dt2 = datetime.now(tz)
+    diff2 = tz.utcoffset(dt2) - tz.dst(dt2)
+    assert diff1 == diff2
+
+
+TESTCASES_MINUTESFROMUTC_TZNAME = [
+
+    # Testcases for MinutesFromUTC.tzname()
+
+    # Each list item is a testcase tuple with these items:
+    # * desc: Short testcase description.
+    # * kwargs: Keyword arguments for the test function:
+    #   * offset: offset argument for MinutesFromUTC().
+    #   * dt: datetime object as input for tzname().
+    #   * exp_tzname: Expected return value of tzname().
+    # * exp_exc_types: Expected exception type(s), or None.
+    # * exp_warn_types: Expected warning type(s), or None.
+    # * condition: Boolean condition for testcase to run, or 'pdb' for debugger
+
+    (
+        "Test with UTC offset 0 and datetime without tzinfo",
+        dict(
+            offset=0,
+            dt=datetime.now(),
+            exp_tzname='00:00',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 30 and datetime without tzinfo",
+        dict(
+            offset=30,
+            dt=datetime.now(),
+            exp_tzname='00:30',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset -30 and datetime without tzinfo",
+        dict(
+            offset=-30,
+            dt=datetime.now(),
+            exp_tzname='-00:30',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 60 and datetime without tzinfo",
+        dict(
+            offset=60,
+            dt=datetime.now(),
+            exp_tzname='01:00',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 719 and datetime without tzinfo",
+        dict(
+            offset=719,
+            dt=datetime.now(),
+            exp_tzname='11:59',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 720 and datetime without tzinfo",
+        dict(
+            offset=720,
+            dt=datetime.now(),
+            exp_tzname='12:00',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 790 and datetime without tzinfo",
+        dict(
+            offset=790,
+            dt=datetime.now(),
+            exp_tzname='13:10',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 999 and datetime without tzinfo",
+        dict(
+            offset=999,
+            dt=datetime.now(),
+            exp_tzname='16:39',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset -999 and datetime without tzinfo",
+        dict(
+            offset=-999,
+            dt=datetime.now(),
+            exp_tzname='-16:39',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset 150 and datetime with tzinfo for EST (ignored)",
+        dict(
+            offset=150,
+            dt=datetime.now(pytz.timezone('US/Eastern')),
+            exp_tzname='02:30',
+        ),
+        None, None, True
+    ),
+    (
+        "Test with UTC offset -150 and datetime with tzinfo for EDT (ignored)",
+        dict(
+            offset=-150,
+            dt=pytz.timezone('US/Eastern').
+            localize(datetime.now(), is_dst=True),
+            exp_tzname='-02:30',
+        ),
+        None, None, True
+    ),
+]
+
+
+@pytest.mark.parametrize(
+    "desc, kwargs, exp_exc_types, exp_warn_types, condition",
+    TESTCASES_MINUTESFROMUTC_TZNAME)
+@simplified_test_function
+def test_MinutesFromUTC_tzname(testcase, offset, dt, exp_tzname):
+    """
+    Test function for MinutesFromUTC.tzname().
+    """
+
+    tz = MinutesFromUTC(offset)
+
+    # The code to be tested
+    act_tzname = tz.tzname(dt)
+
+    # Ensure that exceptions raised in the remainder of this function
+    # are not mistaken as expected exceptions
+    assert testcase.exp_exc_types is None
+
+    assert act_tzname == exp_tzname
 
 
 TESTCASES_CIMTYPE = [

